src/python/pants/backend/java/compile/javac.py

```diff
@@ -5,61 +5,168 @@
 
 import logging
 from dataclasses import dataclass
+from itertools import chain
+from typing import Tuple
 
 from pants.backend.java.target_types import JavaSources
+from pants.build_graph.address import Address
 from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Addresses
+from pants.engine.collection import Collection
 from pants.engine.fs import EMPTY_DIGEST, AddPrefix, Digest, MergeDigests, RemovePrefix
 from pants.engine.process import BashBinary, Process, ProcessResult
 from pants.engine.rules import Get, MultiGet, collect_rules, rule
-from pants.engine.target import Dependencies, DependenciesRequest, Sources, Target, Targets
+from pants.engine.target import (
+    CoarsenedTargets,
+    Dependencies,
+    DependenciesRequest,
+    Sources,
+    Targets,
+)
 from pants.jvm.resolve.coursier_fetch import (
     CoursierLockfileForTargetRequest,
     CoursierResolvedLockfile,
     MaterializedClasspath,
     MaterializedClasspathRequest,
 )
 from pants.jvm.resolve.coursier_setup import Coursier
+from pants.util.frozendict import FrozenDict
 from pants.util.logging import LogLevel
 
 logger = logging.getLogger(__name__)
 
 
 @dataclass(frozen=True)
 class CompileJavaSourceRequest:
-    target: Target
+    targets: CoarsenedTargets
 
 
 @dataclass(frozen=True)
 class CompiledClassfiles:
     digest: Digest
 
 
+class CoarsenedComponents(Collection[CoarsenedTargets]):
+    pass
+
+
+def canonical_addresses_for_coarsened_targets(
+    coarsened_targets: CoarsenedTargets,
+) -> Tuple[Address, ...]:
+    """Return the unique, sorted addresses for a CoarsenedTargets collection.
+
+    This is useful as a canonical ID (e.g. dictionary key) for CoarsenedTargets.
+
+    TODO: Maybe this should be a memoized property of CoarsenedTargets.
+    """
+    return tuple(sorted(frozenset(t.address for t in coarsened_targets)))
+
+
+@rule
+async def coarsened_components(addresses: Addresses) -> CoarsenedComponents:
+    """Returns a deduplicated collection of CoarsenedTargets computed from the input 'addresses'.
```

Comment:
This is a more efficient API for #12251 to expose, yea. All of the naming likely needs work too, and `CoarsenedTargets` could itself probably actually expose the merged/deduped dependencies of each set of targets. I'll tackle that this week to see what you can simplify here.